core/src/main/java/io/grpc/MethodDescriptor.java

```diff
@@ -53,6 +54,81 @@
   private final Marshaller<RequestT> requestMarshaller;
   private final Marshaller<ResponseT> responseMarshaller;
 
+  /**
+   * The call type of a method.
+   */
+  public enum MethodType {
+    /**
+     * One request message followed by one response message.
+     */
+    UNARY,
+
+    /**
+     * Zero or more request messages followed by one response message.
+     */
+    CLIENT_STREAMING,
+
+    /**
+     * One request message followed by zero or more response messages.
+     */
+    SERVER_STREAMING,
+
+    /**
+     * Zero or more request and response messages arbitrarily interleaved in time.
+     */
+    DUPLEX_STREAMING,
+
+    /**
+     * Cardinality and temporal relationships are not known. Implementations should not make
+     * buffering assumptions and should largely treat the same as {@link #DUPLEX_STREAMING}.
+     */
+    UNKNOWN;
+
+    /**
+     * Returns {@code true} if the client will immediately send one request message to the server
+     * after calling {@link ClientCall#start(io.grpc.ClientCall.Listener, io.grpc.Metadata.Headers)}
+     * and then immediately half-close the stream by calling {@link io.grpc.ClientCall#halfClose()}.
+     */
+    public final boolean clientSendsOneMessage() {
+      return this == UNARY || this == SERVER_STREAMING;
+    }
+
+    /**
+     * Returns {@code true} if the server will immediately send one response message to the client
+     * upon receipt of {@link io.grpc.ServerCall.Listener#onHalfClose()} and then immediately
+     * close the stream by calling {@link ServerCall#close(Status, io.grpc.Metadata.Trailers)}.
+     */
+    public final boolean serverSendsOneMessage() {
+      return this == UNARY || this == CLIENT_STREAMING;
+    }
+  }
+
+  /**
+   * Creates a new {@code MethodDescriptor}.
+   *
+   * @param type the call type of this method
+   * @param serviceName the fully qualified service name
+   * @param methodName the short-form method name
+   * @param requestMarshaller the marshaller used to encode and decode requests
+   * @param responseMarshaller the marshaller used to encode and decode responses
+   */
+  public static <RequestT, ResponseT> MethodDescriptor<RequestT, ResponseT> create(
+      MethodType type, String serviceName, String methodName,
+      Marshaller<RequestT> requestMarshaller,
+      Marshaller<ResponseT> responseMarshaller) {
+    return new MethodDescriptor<RequestT, ResponseT>(
```

Comment:
Done. Also renamed the ``name`` field to ``fullMethodName``, as well as the getter.